renderdoc/driver/metal/metal_core.cpp

```diff
@@ -53,3 +56,554 @@ WriteSerialiser &WrappedMTLDevice::GetThreadSerialiser()
+}
+
+void WrappedMTLDevice::CaptureCmdBufCommit(MetalResourceRecord *record)
+{
+  SCOPED_LOCK(m_CaptureCommandBuffersLock);
+  if((record->cmdInfo->flags & MetalCmdBufferStatus::Enqueued) == MetalCmdBufferStatus::NoFlags)
+    CaptureCmdBufEnqueue(record);
+
+  RDCASSERT(MetalCmdBufferStatus::NoFlags ==
+            (record->cmdInfo->flags & MetalCmdBufferStatus::Committed));
+  record->cmdInfo->flags |= MetalCmdBufferStatus::Committed;
+
+  RDCDEBUG("Commit CommandBufferRecord %s", ToStr(record->GetResourceID()).c_str());
+  size_t countSubmitted = 0;
+  for(MetalResourceRecord *record : m_CaptureCommandBuffersEnqueued)
+  {
+    if(record->cmdInfo->flags & MetalCmdBufferStatus::Committed)
+    {
+      RDCASSERT(MetalCmdBufferStatus::NoFlags ==
+                (record->cmdInfo->flags & MetalCmdBufferStatus::Submitted));
+      record->cmdInfo->flags |= MetalCmdBufferStatus::Submitted;
+      ++countSubmitted;
+      CaptureCmdBufSubmit(record);
+      RDCDEBUG("Submit CommandBufferRecord %s", ToStr(record->GetResourceID()).c_str());
+      continue;
+    }
+    break;
+  };
+  m_CaptureCommandBuffersEnqueued.erase(0, countSubmitted);
+}
+
+void WrappedMTLDevice::CaptureCmdBufEnqueue(MetalResourceRecord *record)
+{
+  SCOPED_LOCK(m_CaptureCommandBuffersLock);
+  RDCASSERT(MetalCmdBufferStatus::NoFlags ==
+            (record->cmdInfo->flags & MetalCmdBufferStatus::Enqueued));
+  record->cmdInfo->flags |= MetalCmdBufferStatus::Enqueued;
+  record->AddRef();
+  m_CaptureCommandBuffersEnqueued.push_back(record);
+
+  RDCDEBUG("Enqueing CommandBufferRecord %s %d", ToStr(record->GetResourceID()).c_str(),
+           m_CaptureCommandBuffersEnqueued.count());
+}
+
+void WrappedMTLDevice::AdvanceFrame()
+{
+  if(IsBackgroundCapturing(m_State))
+    RenderDoc::Inst().Tick();
+
+  m_FrameCounter++;    // first present becomes frame #1, this function is at the end of the frame
+}
+
+void WrappedMTLDevice::FirstFrame()
+{
+  // if we have to capture the first frame, begin capturing immediately
+  if(IsBackgroundCapturing(m_State) && RenderDoc::Inst().ShouldTriggerCapture(0))
+  {
+    RenderDoc::Inst().StartFrameCapture(DeviceOwnedWindow(this, NULL));
+
+    m_AppControlledCapture = false;
+    m_CapturedFrames.back().frameNumber = 0;
+  }
+}
+
+void WrappedMTLDevice::Present(MetalResourceRecord *record)
+{
+  WrappedMTLTexture *backBuffer = record->cmdInfo->backBuffer;
+  {
+    SCOPED_LOCK(m_CapturePotentialBackBuffersLock);
+    if(m_CapturePotentialBackBuffers.count(backBuffer) == 0)
+    {
+      RDCERR("Capture ignoring Present called on unknown backbuffer");
+      return;
+    }
+  }
+
+  CA::MetalLayer *outputLayer = record->cmdInfo->outputLayer;
+  DeviceOwnedWindow devWnd(this, outputLayer);
+  {
+    SCOPED_LOCK(m_CaptureOutputLayersLock);
+    if(m_CaptureOutputLayers.count(outputLayer) == 0)
+    {
+      m_CaptureOutputLayers.insert(outputLayer);
+      RenderDoc::Inst().AddFrameCapturer(devWnd, m_Capturer);
```

Comment:
Adding frame capturers like this in present is _extremely_ undesirable and only happens on OpenGL because that API is utterly terrible and its window management is even worse. It also makes it much harder to manage when frame capturers are removed - on GL this requires a soft timeout system of "N seconds since last present", otherwise when windows are created and destroyed .

I know the layers are somewhat outside the metal API, without a clear corresponding API object like a swapchain. I think a better compromise would be to add the frame capturer in the `nextDrawable` call, with similar tracking to know if the layer has been registered already or not. You'd also need some way to know when the layer is destroyed to remove the capturer but perhaps there is a way to get a message or add a hook for the object being destroyed?